components/renku_data_services/user_preferences/orm.py

```diff
@@ -0,0 +1,44 @@
+"""SQLAlchemy schemas for the user preferences database."""
+from typing import Any
+
+from sqlalchemy import JSON, MetaData, String
+from sqlalchemy.dialects.postgresql import JSONB
+from sqlalchemy.orm import DeclarativeBase, Mapped, MappedAsDataclass, mapped_column
+
+from renku_data_services.user_preferences import models
+
+JSONVariant = JSON().with_variant(JSONB(), "postgresql")
+
+metadata_obj = MetaData(schema="user_preferences")  # Has to match alembic ini section name
+
+
+class BaseORM(MappedAsDataclass, DeclarativeBase):
+    """Base class for all ORM classes."""
+
+    metadata = metadata_obj
+
+
+class UserPreferencesORM(BaseORM):
+    """Stored user preferences."""
+
+    __tablename__ = "user_preferences"
+
+    user_id: Mapped[str] = mapped_column("user_id", String(), primary_key=True)
+    """Id of the user."""
+
+    pinned_projects: Mapped[dict[str, Any]] = mapped_column("pinned_projects", JSONVariant)
+    """Pinned projects."""
```

Comment:
Just add a regular int-based `ID` key as the primary key. Make the `user_id` field an index and make it unique as well.

Reply:
Also shouldn't pinned_projects be `Mapped[List[str]]`? I assume this will always be a list of strings in json format.

Reply:
> Just add a regular int-based `ID` key as the primary key. Make the `user_id` field an index and make it unique as well.

Is there a reason for this? I personally think int sequences as primary key are bad because they expose too much data to attackers.

Reply:
storage already uses ULID and we should use that everywhere

Reply:
I dont think you expose any IDs anywhere @leafty. And I agree that we should not use the database IDs in the API.

My comment was to keep the `user_id` and if needed expose that in your API. But, you should not use that as the primary key in the table. Just add a simple `id` field that postgres can autoincrement and and that we will rarely touch directly - but postgres will be happy. And if we have to change the `user_id` field in some way we dont have to worry about dropping the primary key in the table or messing with it and messing up the whole table as a result.

Reply:
> But, you should not use that as the primary key in the table.

Is there a compelling argument here? These are unique within Renku and user preferences will have a 1:1 mapping to user instances.